chatkit-core/src/test/kotlin/com/pusher/chatkit/users/UserSubscriptionDispatcherTest.kt

```diff
@@ -0,0 +1,157 @@
+            )
+            userSubscriptionDispatcher.onEvent(event)
+
+            it("then JoinedRoom is dispatched") {
+                verify(exactly = 1) { dispatcher(JoinedRoom(
+                    room = joinedRoomApiTypeMapper.toRoomInternalType(event.room),
+                    unreadCount = 1
+                )) }
+            }
+        }
+
+        describe("when I receive a RemovedFromRoomEvent") {
+            val event = UserSubscriptionEvent.RemovedFromRoomEvent("room1")
+            userSubscriptionDispatcher.onEvent(event)
+
+            it("then the dispatcher will send a LeftRoom action") {
+                verify(exactly = 1) { dispatcher(LeftRoom(roomId = event.roomId)) }
+            }
+        }
+
+        describe("when I receive a RoomDeletedEvent") {
+            val event = UserSubscriptionEvent.RoomDeletedEvent("room1")
+            userSubscriptionDispatcher.onEvent(event)
+
+            it("then DeleteRoom is dispatched") {
+                verify(exactly = 1) { dispatcher(RoomDeleted(roomId = event.roomId)) }
+            }
+        }
+
+        describe("when I receive a RoomUpdatedEvent") {
+            val event = UserSubscriptionEvent.RoomUpdatedEvent(
+                room = simpleJoinedRoomApiType
+            )
+            userSubscriptionDispatcher.onEvent(event)
+
+            it("then UpdateRoom is dispatched") {
+                verify(exactly = 1) { dispatcher(RoomUpdated(
+                    room = joinedRoomApiTypeMapper.toRoomInternalType(event.room)
+                )) }
+            }
+        }
+    }
+
+    describe("given existing state") {
+        val dispatcher = mockk<Dispatcher>(relaxed = true)
+        val differ = mockk<JoinedRoomsStateDiffer>(relaxed = true)
+        val dateApiTypeMapper = DateApiTypeMapper()
+        val joinedRoomApiTypeMapper = JoinedRoomApiTypeMapper(dateApiTypeMapper)
+        val userSubscriptionDispatcher = UserSubscriptionDispatcher(
+            joinedRoomApiTypeMapper = joinedRoomApiTypeMapper,
+            joinedRoomsStateDiffer = differ,
+            dispatcher = dispatcher
+        )
+
+        every { differ.stateExists() } returns true
+
+        describe("when a new InitialState event is received") {
+            val event = UserSubscriptionEvent.InitialState(
+                currentUser = simpleUser,
+                rooms = listOf(simpleJoinedRoomApiType),
+                readStates = listOf(RoomReadStateApiType("id1", 1, null)),
+                memberships = listOf(RoomMembershipApiType("id1", listOf()))
+            )
+            userSubscriptionDispatcher.onEvent(event)
+
+            it("then the differ toActions will be called") {
+                verify(exactly = 1) {
+                    differ.toActions(
+                        joinedRoomApiTypeMapper.toRoomInternalTypes(event.rooms),
+                        joinedRoomApiTypeMapper.toUnreadCounts(event.readStates))
+                }
```

Comment:
we agreed earlier to have an `IntegrationTest` (using a real Differ) that is more useful than just testing that a function has been called, instead of this test. essentially for the dispatcher we should test what the dispatcher dispatches

alternatively we could also mock differ's toActions to return different actions (with some type repetitions) and check that the dispatcher dispatches the expected actions it should get from the differ

I will be happy with either but the integration test feels better to me, also easy set up and more real test

Reply:
Hmm. I remember agreeing that this test was valuable as the point of `UserSubscriptionTest` is to ensure the right dispatcher methods are called - and this is a valid method that can be called if we want 100% test coverage on this file this was important to cover too.

I remember we did also agree on including an IntegrationTest - and you're right, that's currently missing so I'll work on adding that.

Reply:
Have now added an IntegrationTest and ensured the relevant pathways are covered 😎 The scenarios were copied from the [v1 RoomStoreApplyingNewInitialStatesTest](https://github.com/pusher/chatkit-android/blob/a88be17ef8d1b452a18b55f882c25d440a19b5ab/chatkit-core/src/test/kotlin/com/pusher/chatkit/rooms/RoomStoreApplyingNewInitialStatesTest.kt). However, we're not currently looking at the memberships or readstates yet so this test is expected to be updated over time.

Reply:
awesome, thanks Danielle! I just had a quick look and just provided some minor feedback for future reference. I would have some other minor feedback around naming but there's no much point raising it as we will move that test to the "late" (repository level) differ test when that logic lands there. Likewise as the current "early" differ will be gone there's no point for a detailed review now but it will be best to save that time to do it when that logic lands in the place which has been agreed today to have that responsibility.